/*
 * Copyright 2022 Google LLC.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Framework native row-wise format for the examples.
//
// The Example proto is (currently) only used for worker network communication
// i.e. the users don't feed data using this proto. This proto aims to be
// efficient to handle.

syntax = "proto2";

package yggdrasil_decision_forests.dataset.proto;

option cc_enable_arenas = true;

// One Example (also called observation/record/example/sample).
message Example {
  // Value for multi-dimensional categorical attributes.
  message CategoricalVector {
    repeated int32 values = 1 [packed = true];
  }
  // Value for multi-dimensional numerical attributes.
  message NumericalVector {
    repeated float values = 1 [packed = true];
  }

  // Attribute value.
  message Attribute {
    oneof type {
      bool boolean = 1;
      float numerical = 2;
      int32 categorical = 3;
      string text = 4;
      CategoricalVector categorical_list = 5;
      CategoricalVector categorical_set = 6;
      NumericalVector numerical_list = 7;
      NumericalVector numerical_set = 8;
      // Note: This value will be loaded as a "DiscretizedNumericalIndex =
      // uint16".
      int32 discretized_numerical = 9;
      uint64 hash = 10;
    }
  }
  // Attribute values indexed by the attribute index defined in the dataspec.
  repeated Attribute attributes = 1;
  // Example index.
  optional int64 example_idx = 2;
}
